### ISSUE TITLE
Apollo Server That's Running Inside of Docker can't Connect With Redis Cache on `localhost` 

### ISSUE THREAD

**Issue by Montana** [docker/cli#4865]
### Description

Hi all, 

For some context, I'm trying to establish a subscription server using Apollo Graph 4.0 on AWS Fargate, leveraging the AWS Redis Cache Service for performance enhancement. The strategy involves setting up the system locally before transitioning to the AWS environment. Here's how I've progressed on my local setup:

- Redis Installation: Successfully installed Redis on my local machine, accessible via localhost:6379.
- Redis Verification: I tested the Redis server by pinging it. Receiving a 'pong' response confirmed its operational status.
- Apollo Server Setup: I then proceeded to construct and launch an Apollo Server 4.0 instance. This server, once initiated from the terminal, connects to Redis without any issues, indicating a successful setup thus far.
Dockerization: Moving forward, I encapsulated the Apollo Graph server within a Docker container. 

The container was created using the command: `docker build -t apollo-server .` Subsequently, I attempted to run the Docker image using various commands:

```bash
docker run --network host apollo-server
docker run --network host -p 6379:6379 apollo-server
docker run --network host -p 4500:4500 apollo-server
```
However, upon initiation, the Apollo Graph server within the Docker container encountered an error: `Error: connect ECONNREFUSED 127.0.0.1:6379`, indicating an inability to establish a connection with the Redis server.

### Reproduce

```bash
docker run --network host apollo-server
docker run --network host -p 6379:6379 apollo-server
docker run --network host -p 4500:4500 apollo-server
```

### Expected behavior

Connection being made.

### docker version

```bash
Client: Docker Engine - Community
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.18.3
 Git commit:        100c70180f
 Built:             Mon Jun  6 21:36:39 2022
 OS/Arch:           darwin/arm64
 Context:           desktop-linux
 Experimental:      true

Server: Docker Desktop 4.27.1 (136059)
 Engine:
  Version:          25.0.2
  API version:      1.44 (minimum version 1.24)
  Go version:       go1.21.6
  Git commit:       fce6e0c
  Built:            Thu Feb  1 00:23:21 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.28
  GitCommit:        ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    desktop-linux
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.12.1-desktop.4)
  compose: Docker Compose (Docker Inc., v2.24.3-desktop.1)
  debug: Get a shell into any image or container. (Docker Inc., 0.0.22)
  dev: Docker Dev Environments (Docker Inc., v0.1.0)
  extension: Manages Docker extensions (Docker Inc., v0.2.21)
  feedback: Provide feedback, right in your terminal! (Docker Inc., v1.0.4)
  init: Creates Docker-related starter files for your project (Docker Inc., v1.0.0)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scout: Docker Scout (Docker Inc., v1.3.0)
WARNING: Plugin "/Users/montana/.docker/cli-plugins/docker-scan" is not valid: failed to fetch metadata: fork/exec /Users/montana/.docker/cli-plugins/docker-scan: no such file or directory

Server:
 Containers: 22
  Running: 8
  Paused: 0
  Stopped: 14
 Images: 31
 Server Version: 25.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: active
  NodeID: 16qlmclkmig8ktsmg1hzyxe6e
  Is Manager: true
  ClusterID: 4bnpcgiakdt7ienzfbtnq672t
  Managers: 1
  Nodes: 1
  Default Address Pool: 10.0.0.0/8
  SubnetSize: 24
  Data Path Port: 4789
  Orchestration:
   Task History Retention Limit: 5
  Raft:
   Snapshot Interval: 10000
   Number of Old Snapshots to Retain: 0
   Heartbeat Tick: 1
   Election Tick: 10
  Dispatcher:
   Heartbeat Period: 5 seconds
  CA Configuration:
   Expiry Duration: 3 months
   Force Rotate: 0
  Autolock Managers: false
  Root Rotation In Progress: false
  Node Address: 192.168.65.3
  Manager Addresses:
   192.168.65.3:2377
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: ae07eda36dd25f8a1b98dfbf587313b99c0190bb
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  seccomp
   Profile: unconfined
  cgroupns
 Kernel Version: 6.6.12-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: aarch64
 CPUs: 10
 Total Memory: 7.657GiB
 Name: docker-desktop
 ID: 72fa8517-08c2-4245-80c1-dd706003f143
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: http.docker.internal:3128
 HTTPS Proxy: http.docker.internal:3128
 No Proxy: hubproxy.docker.internal
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  hubproxy.docker.internal:5555
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
> Redis Installation: Successfully installed Redis on my local machine, accessible via localhost:6379.


From your output, it looks like you are running Docker Desktop. In that case, Linux containers (which requires a Linux kernel) run inside a VM. Using `host` networking will make those containers share the networking namespace on the host that the daemon runs on, which in that case is the Linux VM, and `localhost` is the `localhost` of the VM.

On docker desktop you can use `host.docker.internal` to connect to services running on your (macOS) host; see https://docs.docker.com/desktop/faqs/general/#how-do-i-connect-from-a-container-to-a-service-on-the-host

Does that help?

**Comment by Montana** [comment]
Hi @thaJeztah, 

I ended up solving the issue at hand, and I'll give an RCA for folks who might have it in the future. 

Docker and AWS Fargate, containers do not communicate via localhost (`127.0.0.1`) for inter-container communication unless they are in the same task definition and using the `awsvpc` network mode. Even then, as I understand it you would typically use the service discovery mechanism rather than `localhost`. 

I enforced the containers to use the `awsvpc` networking mode, which provides each task with its own elastic network interface, which allowed me to apply security groups. Then with your suggestion @thaJeztah using `host.docker.internal` to connect outside services (such as Fargate) worked.
